### PR TITLE
Enable running tests for cached images

### PR DESCRIPTION
**Pull request by PriyaModali**
Fixes: #3879

**Description**
Enable running tests for cached images.

Original behavior:
Tests were being skipped for cached images.

Updated behavior:
The tests are always run unless `--skip-tests` flag is set
Log an info message when the `--skip-tests` flag is set

How:
Moved the code that runs the tests to outside the caching logic
Tests are now run always regardless of the caching logic

Before this fix the tests were being skipped when artifacts are found in the cache. But this PR changes that so that tests are run for both cached images and newly build artifacts and log a message when skip-tests is set to true

**User facing changes**
Tests will always run (except when skip-tests flag is explicitly set)

### CODE REVIEW

File: pkg/skaffold/runner/build_deploy.go [hunk 1]
```diff
@@ -73,18 +73,20 @@ func (r *SkaffoldRunner) BuildAndTest(ctx context.Context, out io.Writer, artifa
...
 		return nil, err
 	}
 
+	if !r.runCtx.SkipTests() {
+		if err = r.tester.Test(ctx, out, bRes); err != nil {
+			return nil, err
+		}
+	} else {
+		logrus.Warnf("Skip tests flag (skip-tests) is set to %t. Tests won't run.", (r.runCtx.SkipTests()))
+	}
```
Comment (gsquared94) [inline pkg/skaffold/runner/build_deploy.go:88]: nit: invert `if-else`
```suggestion
	if r.runCtx.SkipTests() {
		logrus.Warnf("Skip tests flag (skip-tests) is set to %t. Tests won't run.", (r.runCtx.SkipTests()))
	} else {
		if err = r.tester.Test(ctx, out, bRes); err != nil {
			return nil, err
		}
	}
```

File: pkg/skaffold/runner/build_deploy.go [hunk 2]
```diff
@@ -73,18 +73,20 @@ func (r *SkaffoldRunner) BuildAndTest(ctx context.Context, out io.Writer, artifa
...
 	if err != nil {
 		return nil, err
 	}
 
+	if !r.runCtx.SkipTests() {
+		if err = r.tester.Test(ctx, out, bRes); err != nil {
+			return nil, err
+		}
+	} else {
+		logrus.Warnf("Skip tests flag (skip-tests) is set to %t. Tests won't run.", (r.runCtx.SkipTests()))
```
Comment (gsquared94) [inline pkg/skaffold/runner/build_deploy.go:87]: `r.runCtx.SkipTests()` is always true here.

```suggestion
		logrus.Warnln("Skip tests flag (skip-tests) is set to true. Tests won't run.")
```

**Comment by tejal29** [comment]
Please fill in the details in the PR description describing the change you did to fix the bug. 
I see you mention in the Pr description "Enable running tests for cached images " however it doesn't describe the "how" part.

Re "and log a meaningful message.", i dont think the message is changed in this PR.

**Comment by tejal29** [comment]
> Please fill in the details in the PR description describing the change you did to fix the bug.
> I see you mention in the Pr description "Enable running tests for cached images " however it doesn't describe the "how" part.
> 
> Re "and log a meaningful message.", i dont think the message is changed in this PR.

Please correct the PR description to mention what is addressed in the PR.

**Comment by tejal29** [comment]
> > Please fill in the details in the PR description describing the change you did to fix the bug.
> > I see you mention in the Pr description "Enable running tests for cached images " however it doesn't describe the "how" part.
> > Re "and log a meaningful message.", i dont think the message is changed in this PR.
> 
> Please correct the PR description to mention what is addressed in the PR.

ah, i see it now. you have added a warning message when `--skip-tests` is present on the command line. 
I think we should drop it. It made sense to show a warning when users did not have the flag and skaffold skipped testing images.
With your change of running tests, i feel showing this warning is redundant. 
WDYT?

**Comment by PriyaModali** [comment]
> > > Please fill in the details in the PR description describing the change you did to fix the bug.
> > > I see you mention in the Pr description "Enable running tests for cached images " however it doesn't describe the "how" part.
> > > Re "and log a meaningful message.", i dont think the message is changed in this PR.
> > 
> > 
> > Please correct the PR description to mention what is addressed in the PR.
> 
> ah, i see it now. you have added a warning message when `--skip-tests` is present on the command line.
> I think we should drop it. It made sense to show a warning when users did not have the flag and skaffold skipped testing images.
> With your change of running tests for cached artifacts, i feel showing this warning is redundant.
> WDYT?

This change would always warn the user in both cached & non-cached scenarios letting the user know that
- the tests are not run as the skip-tests flag is set 
- their changes have not been tested.

**Comment by tejal29** [comment]
> > > > Please fill in the details in the PR description describing the change you did to fix the bug.
> > > > I see you mention in the Pr description "Enable running tests for cached images " however it doesn't describe the "how" part.
> > > > Re "and log a meaningful message.", i dont think the message is changed in this PR.
> > > 
> > > 
> > > Please correct the PR description to mention what is addressed in the PR.
> > 
> > 
> > ah, i see it now. you have added a warning message when `--skip-tests` is present on the command line.
> > I think we should drop it. It made sense to show a warning when users did not have the flag and skaffold skipped testing images.
> > With your change of running tests for cached artifacts, i feel showing this warning is redundant.
> > WDYT?
> 
> This change would always warn the user in both cached & non-cached scenarios letting the user know that
> 
> * the tests are not run as the skip-tests flag is set
> * their changes have not been tested.

I agree. I feel the warning is redundant since users explicitly used the flag `--skip-tests`

**Review by tejal29 — CHANGES_REQUESTED** [review]
Don't warn if skip test flag is present. 

Please address comment https://github.com/GoogleContainerTools/skaffold/pull/5013#issuecomment-724878579 to update the description.

File: pkg/skaffold/runner/build_deploy.go [hunk 3]
```diff
@@ -80,7 +80,7 @@ func (r *SkaffoldRunner) BuildAndTest(ctx context.Context, out io.Writer, artifa
 	}
 
 	if r.runCtx.SkipTests() {
-		logrus.Warnln("Skip tests flag (skip-tests) is set to true. Tests won't run.")
```
Comment (tejal29) [inline pkg/skaffold/runner/build_deploy.go]: Lets not print any message. Skaffold already prints a lot of output. The team is rework only printing what is relevant.

**Review by tejal29 — CHANGES_REQUESTED** [review]

File: pkg/skaffold/runner/build_deploy.go (hunk 3 above)
Comment (PriyaModali) [reply pkg/skaffold/runner/build_deploy.go]: Removed the log message.

**Review by tejal29 — APPROVED** [review]

**Pull request merged**